File: sdk/python/gnmi/tests/test_utils.py
#  ----------------------------------------------------------------
# Copyright 2016 Cisco Systems
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------

"""test_utils.py

Utility function for test cases.
"""
import re
import unittest
import logging

from argparse import ArgumentParser

import sys
if sys.version_info > (3,):
    from urllib.parse import urlparse
else:
    from urlparse import urlparse

from ydk.entity_utils import get_data_node_from_entity

class EmptyTest(unittest.TestCase):
    @classmethod
    def setUpClass(self):
        pass
    def runTest(self):
        pass

def assert_with_error(pattern, ErrorClass):
    def assert_with_pattern(func):
        def helper(self, *args, **kwargs):
            try:
                func(self)
            except ErrorClass as error:
                res = re.match(pattern, error.message.strip())
                self.assertEqual(res is not None, True)
        return helper
    return assert_with_pattern

def enable_logging(level):
    log = logging.getLogger('ydk')
    log.setLevel(level)
    handler = logging.StreamHandler()
    formatter = logging.Formatter(("%(asctime)s - %(name)s - %(levelname)s - %(message)s"))
    handler.setFormatter(formatter)
    log.addHandler(handler)

def datanode_to_str(dn, indent = ''):
    try:
        s = dn.get_schema_node().get_statement()
        if s.keyword == "leaf" or s.keyword == "leaf-list" or s.keyword == "anyxml":
            out = indent + "<" + s.arg + ">" + dn.get_value() + "</" + s.arg + ">\n"
        else:
            out = indent + "<" + s.arg + ">\n"
            child_indent = indent + "  "
            for child in dn.get_children():
                out += datanode_to_str(child, child_indent)
            out += indent + "</" + s.arg + ">\n"
        return out
    except YCoreError as ex:
        print(ex.message)

def print_data_node(dn):
    try:
        print("\n=====>  Printing DataNode: '{}'".format(dn.get_path()))
        print(datanode_to_str(dn))
    except YPYCoreError as ex:
        print(ex.message)

def print_entity(entity, root_schema):
    dn = get_data_node_from_entity( entity, root_schema);
    print_data_node(dn)

def entity_to_string(entity, root_schema):
    dn = get_data_node_from_entity( entity, root_schema);
    return datanode_to_str(dn)

class ParametrizedTestCase(unittest.TestCase):
    """ TestCase classes that want to be parametrized should
        inherit from this class.
    """
    def __init__(self, methodName='runTest'):
        super(ParametrizedTestCase, self).__init__(methodName)

    @staticmethod
    def parametrize(testcase_klass, device):
        """ Create a suite containing all tests taken from the given
            subclass, passing them the parameter 'param'.
        """
        testloader = unittest.TestLoader()
        testcase_klass.hostname = device.hostname
        testcase_klass.username = device.username
        testcase_klass.password = device.password
        testcase_klass.port = device.port
        testnames = testloader.getTestCaseNames(testcase_klass)
        suite = unittest.TestSuite()
        for name in testnames:
            suite.addTest(testcase_klass(name))
        return suite

def get_device_info():
    parser = ArgumentParser()
    parser.add_argument("-v", "--verbose", help="print debugging messages", action="store_true")
    parser.add_argument("device", nargs='?',
                        help="gNMI server credentials in format: 'ssh://user:password@host:port'")

    args = parser.parse_args()
    if not args.device:
        args.device = "ssh://admin:admin@127.0.0.1:50051"
    device = urlparse(args.device)
    return device

def get_local_repo_dir():
    import os
    path = os.path.dirname(os.path.abspath(__file__))
    path = os.path.join(path, '../../../cpp/core/tests/models')
    return path

File: sdk/cpp/gnmi/tests/test_utils.cpp
/// YANG Development Kit
// Copyright 2016 Cisco Systems. All rights reserved
//
////////////////////////////////////////////////////////////////
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
//////////////////////////////////////////////////////////////////

#include <ydk/gnmi_service.hpp>
#include <ydk/entity_data_node_walker.hpp>

#include <ydk/filters.hpp>

#include "test_utils.hpp"

using namespace std;

using namespace ydk;
using namespace path;
using namespace ydktest;

std::string print_tree(ydk::path::DataNode* dn, const std::string& indent)
{
  ostringstream buffer;
  try {
    ydk::path::Statement s = dn->get_schema_node().get_statement();
    if(s.keyword == "leaf" || s.keyword == "leaf-list" || s.keyword == "anyxml") {
        auto val = dn->get_value();
        buffer << indent << "<" << s.arg << ">" << val << "</" << s.arg << ">" << std::endl;
    } else {
        std::string child_indent{indent};
        child_indent+="  ";
        buffer << indent << "<" << s.arg << ">" << std::endl;
        for(auto c : dn->get_children())
        	buffer << print_tree(c.get(), child_indent);
        buffer << indent << "</" << s.arg << ">" << std::endl;
    }
  }
  catch (ydk::path::YCoreError &ex) {
	  cout << ex.what() << endl;
  }
  return buffer.str();
}

void print_data_node(shared_ptr<ydk::path::DataNode> dn)
{
  try {
    cout << "\n=====>  Printing DataNode: '" << dn->get_path() << "'" << endl;
    cout << print_tree(dn.get(), " ");
  }
  catch (ydk::path::YCoreError &ex) {
    cout << ex.what() << endl;
  }
}

void print_entity(shared_ptr<ydk::Entity> entity, ydk::path::RootSchemaNode& root)
{
    ydk::path::DataNode& dn = get_data_node_from_entity( *entity, root);
    ydk::path::Statement s = dn.get_schema_node().get_statement();
    cout << "\n=====>  Printing DataNode: '" << s.arg << "'" << endl;
    cout << print_tree( &dn, " ");
}

string entity2string(shared_ptr<ydk::Entity> entity, ydk::path::RootSchemaNode& root)
{
    ydk::path::DataNode& dn = get_data_node_from_entity( *entity, root);
    ydk::path::Statement s = dn.get_schema_node().get_statement();
    return print_tree( &dn, " ");
}

void config_bgp(openconfig_bgp::Bgp& bgp)
{
    bgp.global->config->as = 65172;

    auto neighbor = make_shared<openconfig_bgp::Bgp::Neighbors::Neighbor>();
    neighbor->neighbor_address = "172.16.255.2";
    neighbor->config->neighbor_address = "172.16.255.2";
    neighbor->config->peer_as = 65172;

    bgp.neighbors->neighbor.append(neighbor);
}

void build_bgp_config(gNMIServiceProvider& provider)
{
    openconfig_bgp::Bgp bgp{};
    bgp.yfilter = YFilter::replace;
    config_bgp(bgp);

    gNMIService gs{};
    gs.set(provider, bgp);
}

void build_int_config(gNMIServiceProvider& provider)
{
    auto ifc = openconfig_interfaces::Interfaces::Interface();
    ifc.name = "Loopback10";
    ifc.config->name = "Loopback10";
    ifc.config->description = "Test";
    ifc.yfilter = YFilter::replace;

    gNMIService gs{};
    gs.set(provider, ifc);
}

void delete_bgp_config(gNMIServiceProvider& provider)
{
    openconfig_bgp::Bgp bgp{};
    bgp.yfilter = YFilter::delete_;

    gNMIService gs{};
    gs.set(provider, bgp);
}

void delete_int_config(gNMIServiceProvider& provider)
{
    openconfig_interfaces::Interfaces ifcs{};
    ifcs.yfilter = YFilter::delete_;

    gNMIService gs{};
    gs.set(provider, ifcs);
}
